indigo/pipelines/html.py

```diff
@@ -0,0 +1,318 @@
+                    filename = f'img{helper["counter"]}.{file_ext}'
+                    context.attachments.append(ImportAttachment(
+                        filename=filename,
+                        content_type=image.content_type,
+                        file=f,
+                    ))
+                return {'src': 'media/' + filename}
+            except KeyError as e:
+                # raised when the image can't be found in the zip file
+                log.info(f"Image cannot be found in docx file; ignoring", exc_info=e)
+                return {}
+
+        try:
+            result = mammoth.convert_to_html(context.source_file, convert_image=mammoth.images.img_element(stash_image))
+            html = result.value
+        except BadZipFile:
+            raise ValueError("This doesn't seem to be a valid DOCX file.")
+
+        context.html_text = html
+
+
+class ParseHtml(Stage):
+    """ Parse html with lxml.html.
+
+    Reads: context.html_text
+    Writes: context.html
+    """
+    def __call__(self, context):
+        context.html = html.fromstring(context.html_text)
+
+
+class CleanHtml(Stage):
+    """ Clean dangerous HTML.
+
+    Reads: context.html
+    Writes: context.html
+    """
+    cleaner = Cleaner(
+        style=True,
+        inline_style=False,
+        safe_attrs=list(Cleaner.safe_attrs) + ['style']
+    )
+
+    def __call__(self, context):
+        context.html = self.cleaner.clean_html(context.html)
+
+
+class SimplifyHtml(Stage):
+    """ Simplify HTML:
+
+    * strong to b
+    * em to i
+    * remove unused headers, styles, scripts etc.
+    """
+
+    xsl = """
+<xsl:stylesheet version="1.0" xmlns:xsl="http://www.w3.org/1999/XSL/Transform">
+
+  <xsl:output method="html" encoding="utf-8" />
+  <xsl:strip-space elements="p"/>
+
+  <!-- ignore these elements -->
+  <xsl:template match="head|style|script|link" />
+
+  <!-- unwrap these elements -->
+  <xsl:template match="div[not(starts-with(@id, 'sdfootnote'))] | span | font">
+    <xsl:apply-templates />
+  </xsl:template>
+
+  <!-- strong to b -->
+  <xsl:template match="strong">
+    <b><xsl:apply-templates /></b>
+  </xsl:template>
+
+  <!-- em to i -->
+  <xsl:template match="em">
+    <i><xsl:apply-templates /></i>
+  </xsl:template>
+
+  <!-- for most nodes, just copy -->
+  <xsl:template match="@*|node()">
+    <xsl:copy>
+      <xsl:apply-templates select="@*|node()"/>
+    </xsl:copy>
+  </xsl:template>
+
+</xsl:stylesheet>
+"""
+
+    def __call__(self, context):
+        pass
+        xslt = etree.XSLT(etree.fromstring(self.xsl))
+        context.html = html.fromstring(str(xslt(context.html)))
+
+
+class NormaliseHtmlTextWhitespace(Stage):
+    """ Strip and normalise whitespace in HTML text.
+
+    Reads: context.html_text
+    Writes: context.html_text
+    """
+    def __call__(self, context):
+        # &nbsp; to space
+        context.html_text = context.html_text.replace('&nbsp;', ' ')
```

Comment:
Side topic: I'd love for us to keep and somehow support non-breaking spaces at some point